### PR TITLE
Add DeriviedHandlers sample

### PR DESCRIPTION
**Pull request by HaoK**
Its fairly involved to just get a trivial derived cookie authentication handler that takes uses its own derived options from cookie options, and extends CookieAuthentication options.

I figure we should review this initial checkpoint before doing anything real since its already pretty complex...

DerivedCookieHandler.cs has all the interesting stuff

@Tratcher

### CODE REVIEW

**Comment by HaoK** [comment]
Basically, there's a few pain points.

The base CookieAuthenticationHandler class expects `IOptionsMonitor<CookieAuthenticationOptions>`, so I got around this by just passing in a custom one that wraps the real `IOptionsMonitor<DerivedCookieAuthenticationOptions>`

Deriving also breaks the `AddScheme<TOptions, THandler>` sugar since we aren't deriving from `AuthenticationHandler<DerivedCookieOptions>` either.  So I had to reimplement the internal sugar logic in the AddDerivedCookie.

Finally we can't easily reuse the cookie post configure either due to the custom options type...

File: samples/DerivedHandlers/README.md [hunk 1]
```diff
@@ -0,0 +1,9 @@
+AuthSamples.DerivedHandlers
```
Comment (kichalla) [inline samples/DerivedHandlers/README.md]: I feel this doc requires more details. For example, you mention `// Normally Identity handles sign in, but you can do it directly` in `AccountController`. Like mentioning the differences in what by default happens vs. what this sample does probably would be helpful.

Comment (HaoK) [reply samples/DerivedHandlers/README.md]: Do you mean just expanding on the comments?  The cookie samples aren't really focusing on the identity generation part of things, but I guess its probably worth at least pointing to a real doc/topic talking about this

**Comment by Tratcher** [comment]
OIDC and JWT are the other major candidates for derivation. OAuth already has it built in.

**Comment by Tratcher** [comment]
Yeah, this is a pretty terrible experience. Do we have any known scenarios for trying to derive from CookieAuthHandler? E.g. what would they want to customize? Using the events or forking the code look far more practical.

**Comment by HaoK** [comment]
Its really going to be the same for all of them, the only reason I picked cookie was because it already had a dedicated sample in here, so it was the least amount of work to try first.

So far I think the two main scenarios that people try to do are, adding different options which they use to compute the actual auth options (this can be accomplished via options magic as a workaround, which is what we do for AAD and OIDC).  The other is when they want to customize the behavior at a point where there's no event.  That's the main thing that's missing, tweaking the default flow slightly via overriding challenge but yet still being able to call base/reuse some of the flow.

**Comment by Tratcher** [comment]
Derivation is insufficiently granular for tweaking the default behaviors. You end up having to re-implement all of the given action you override (e.g. SignIn). At that point forking is more practical.

**Comment by HaoK** [comment]
So should we scope this sample purely to the 'skinning' an existing auth options?

aka how to do what AAD was doing?  Register MyCookieOptions, and the a sample for how you can configure MyCookieOptions and underneath the covers automatically translate that into an AddCookie() and computing the 'real' cookie options from MyCookieOptions.  

Basically its a sample for Computing options from another options I guess as opposed to derivation at all.

**Comment by HaoK** [comment]
But its kind of hard to argue there's no derivation scenario when we do it for OAuth

**Comment by Tratcher** [comment]
Certainly, the difference is that OAuth was designed for derivation and you override granular, protocol specific APIs like FormatScope, not large sections like AuthenticateAsync. We also placed a lot of restrictions like no custom event types. CookieAuth, JWT, and OIDC would need redesigning if we really wanted to make them derivation friendly and I don't think we have any scenarios that require it.

**Comment by Tratcher** [comment]
Yeah, even in Katana we theorized derivation would be a useful extensiblity model, but we never did the exercise to prove it.

**Comment by HaoK** [comment]
OAuth is already your proof, everything is extensible, except for Cookies/Jwt/Google/Facebook since that work was just never done.  I'm not really pushing for doing that work, but the scenario is clear, its if you want to reuse parts of the logic without copying and pasting, its the same reason why we made everything else reusable.  We just never did the work to refactor/clean up the old katana code, it was always on my todo list, but its never (and probably never will) make it to the top of the todo list.

Events might be easier to use but they only expose exactly those extensbility points, there's no easy way to add new ones or change behavior other than forking the code.

**Comment by HaoK** [comment]
Close I think this isn't useable enuf to show in a sample

**Pull request closed without merging**